devito/operator.py

```diff
@@ -114,63 +119,58 @@ def __init__(self, expressions, **kwargs):
         # Finish instantiation
         super(Operator, self).__init__(self.name, nodes, 'int', parameters, ())
 
-    def _argument_defaults(self, arguments):
-        """
-        Derive all default values from parameters and ensure uniqueness.
-        """
-        default_args = ArgumentMap()
-        for p in self.input:
-            if p.name not in arguments:
-                default_args.update(p.argument_defaults())
-        for p in self.dimensions:
-            if p.name not in arguments and p.is_Sub:
-                default_args.update(p.argument_defaults(default_args))
-        return {k: default_args.reduce(k) for k in default_args if k not in arguments}
-
     def arguments(self, **kwargs):
         """
         Process runtime arguments passed to ``.apply()` and derive
         default values for any remaining arguments.
         """
-        arguments = {}
-        # First, we insert user-provided override
-        for p in self.input + self.dimensions:
-            arguments.update(p.argument_values(**kwargs))
+        # Handle data-carriers (first overrides, then defaults)
+        args = ReducerMap()
+        args.update([p._arg_values(**kwargs) for p in self.input if p.name in kwargs])
+        args.update([p._arg_defaults() for p in self.input if p.name not in args])
+        args = args.reduce_all()
 
-        # Second, derive all remaining default values from parameters
-        arguments.update(self._argument_defaults(arguments))
+        # Handle dimensions (first adjust data-carriers-induced defaults, then overrides)
+        for p in self.dimensions:
+            args.update(p._arg_infers(args, self._dspace[p], **kwargs))
+        for p in self.dimensions:
+            args.update(p._arg_values(**kwargs))
+
```

Comment:
Would switch, same as `input`, first fill `kwargs` then complete with missing remaining ones.

Reply:
why? why should I bother fetching defaults if I'm not gonna use them?

Reply:
OK, hang on, maybe I didn't understand the comment. ``infers`` must go before ``_arg_values``, because the way I implemented the logic here is that ``infers`` works on defautls

Reply:
I mean

```
for p in self.dimensions:
            args.update(p._arg_values(**kwargs))
for p in self.dimensions and not in args:
            args.update(p._arg_infers(args, self._dspace[p], **kwargs))
```

Reply:
No; dimension overrides are fixed values, while inferred ones are meant to be derived from defaults; this made me realize I should probably find a more sensible name for this method as, I agree, it might sound confusing

Reply:
I cannot think of a better name. Any ideas?

Reply:
I'm not sure I understand what you mean.
- overrides => runtime value
- infered => default

It's just number so it doesn't really matter, just was wandering if order in which is done can be consotent with the order for Function
- inputs, then default for the rest

Reply:
right, and no there's a logic behind it, which is why I said the name "infer" may be misleading. With ``infer`` you add/subtract points to avoid OOB accesses; but you don't want to add/subtract points to overrides. An override is fixed, it's given by the user and that's it. So we run ``infer``, which adds/subtracts points to the values generated by processing ``Function``s (``time_m, time_M, x_m, ...``) and then, if we have overrides, we simply replace them with such overrides

Reply:
You don't need to derive defaults for something that will be overridden with a user-provided value in the next step.

Reply:
what's the alternative? guarding with ``ifs`` ?

Reply:
Citing @mloubout from above: 
> first fill kwargs then complete with missing remaining ones.

Reply:
I'm not sure how to explain this after what I wrote above. Feel free to try and switch it, it won't work unless with non-trivial changes (ie adding guards).